Add a pluggable logging backend

atakanarikan opened the pull request:
This PR adds a setting to the library, making it more pluggable to end users.

We have encountered that the database table became too large for our operations, and decided to log our audit events to ElasticSearch instead. With the changes in this pr, it's possible to change the logging backend to whatever the end user prefers, instead of being forced to create a database table for audit logs.

A simple example for kibana is given in the readme.

jheld commented:
Sounds really cool! My team was planning to consider other NoSQL-ish solutions too. We like the flow of easyaudit, but haven't considered a pluggable backend.

I should have time to review it next week.

morlandi commented:
+1 Nice !

jheld commented on README.md:
```diff
@@ -51,14 +51,41 @@ Below are some of the settings you may want to use. These should be defined in y
+    import logging
+    
+    class PythonLoggerBackend:
+        logger = logging.getLogger('your-kibana-logger')
```
Minor improvement (likely just in the future) to be able to specify the logging level per each object type.

Can get around it via a subclass but still would be nice to have some way to specify that without having to subclass and override the class.

I'm not advocating for holding up the merge, but thought it would be a good enhancement.

jheld commented on README.md:
```diff
@@ -51,14 +51,41 @@ Below are some of the settings you may want to use. These should be defined in y
+    class PythonLoggerBackend:
+        logger = logging.getLogger('your-kibana-logger')
+        
+        def request(self, request_info):
```
For consistency (even though it isn't super useful?), we may wish to return the info object back to the caller.

I'm thinking that if we're making this truly pluggable, then the core classes should try to follow a consistent behavior.

With type annotations, we could try to enforce it style-wise, too.

soynatan replied:
@jheld I agree! Can you type your suggestion?

soynatan commented:
Woah, pretty cool feature! Let's get it in master.

jheld commented:
We are still very interested in this PR. The suggested changes are pretty minor overall. Could you plan some time by end of year for this? I might be able to fork the branch and try myself, but that might take some time to get at it.

atakanarikan commented:
Yeah I can definitely address things by the end of the year. I'm a bit busy atm, but things should clear up soon. Thanks for the update!

jheld commented:
Hi @atakanarikan happy new year.

Could you provide an eta around this? If you're unsure, would you mind if (time allowing) I branch off this and merge with the proposed changes?

atakanarikan commented:
Hey guys, thanks for the new year wish, same to you!

Sorry about the delay with this, things have been quite busy with me. I changed my organization and no longer have access to the for for this pr. Therefore I created another PR #124  where I (hopefullly) addressed your comments. I will close this pr, if you have further comments please refer to #124 which isn't forked to an org but to my personal account.

Best wishes!
